lib/python/pyflyby/_interactive.py

```diff
@@ -759,174 +753,65 @@ def get_global_namespaces(ip):
-        if sys.modules.get(pname, Ellipsis) is parent and parent.__name__ == pname:
-            # Add known_imports entries from the database.
-            results.update(known.member_names.get(pname, []))
-            # Get the module handle.  Note that we use ModuleHandle() on the
-            # *name* of the module (``pname``) instead of the module instance
-            # (``parent``).  Using the module instance normally works, but
-            # breaks if the module hackily replaced itself with a pseudo
-            # module (e.g. https://github.com/josiahcarlson/mprop).
-            pmodule = ModuleHandle(pname)
-            # Add importable submodules.
-            results.update([m.name.parts[-1] for m in pmodule.submodules])
-        results = sorted([r for r in results if r.startswith(attrname)])
-        results = ["%s.%s" % (pname, r) for r in results]
-    else:
-        raise AssertionError
-    logger.debug("complete_symbol(%r) => %r", fullname, results)
-    return results
-
+        results = sorted([r for r in results])
+        return results
 
-def _list_members_for_completion(obj, ip):
-    """
-    Enumerate the existing member attributes of an object.
-    This emulates the regular Python/IPython completion items.
+    def keys(self):
+        return list(dict.keys(self)) + self._potential_imports_list
 
-    It does not include not-yet-imported submodules.
 
-    :param obj:
-      Object whose member attributes to enumerate.
-    :rtype:
-      ``list`` of ``str``
-    """
-    if ip is None:
-        words = dir(obj)
-    else:
-        try:
-            limit_to__all__ = ip.Completer.limit_to__all__
-        except AttributeError:
-            limit_to__all__ = False
-        if limit_to__all__ and hasattr(obj, '__all__'):
-            words = getattr(obj, '__all__')
-        elif "IPython.core.error" in sys.modules:
-            from IPython.utils import generics
-            from IPython.utils.dir2 import dir2
-            from IPython.core.error import TryNext
-            words = dir2(obj)
-            try:
-                words = generics.complete_object(obj, words)
-            except TryNext:
-                pass
-        else:
-            words = dir(obj)
-    return [w for w in words if isinstance(w, str)]
+def _auto_import_hook(name: str):
+    logger.debug("_auto_import_hook(%r)", name)
+    ip = _get_ipython_app().shell
+    try:
+        namespaces = get_global_namespaces(ip)
+        namespaces = ScopeStack(namespaces)
+        db = None
+        db = ImportDB.interpret_arg(db, target_filename='.')
+        did_auto_import = auto_import_symbol(name, namespaces, db)
+    except Exception as e:
+        logger.debug("_auto_import_hook preparation error: %r", e)
+        raise e
+    if not did_auto_import:
+        raise ImportError(f"{name} not auto-imported")
+    try:
+        # relies on `auto_import_symbol` auto-importing into [-1] namespace
+        return namespaces[-1][name]
```

Comment:
Having not looked at `auto_import_symbol`, is there a case where this access can fail? Seems like `auto_import_symbol` should guarantee it appends to last namespace or fail if not?

Reply:
If auto-importing in `auto_import_symbol` fails then ought to return `False`. If it does not then it promises to insert the name into the last namespace as per docstring:

https://github.com/deshaw/pyflyby/blob/d46890cd83645a9ae8107bd5447c3d3d9afab13e/lib/python/pyflyby/_autoimp.py#L1854-L1855

The actual implementation does catch any errors:

https://github.com/deshaw/pyflyby/blob/d46890cd83645a9ae8107bd5447c3d3d9afab13e/lib/python/pyflyby/_autoimp.py#L1820-L1841

So if it fails I think this would be an issue to address in `auto_import_symbol` and not in `_auto_import_hook` as  `auto_import_symbol` does not declare that it can raise.